torch/onnx/_type_utils.py

```diff
@@ -108,36 +113,86 @@ def from_name(
     @classmethod
     @_beartype.beartype
     def from_dtype(cls, dtype: torch.dtype) -> JitScalarType:
-        """Convert a torch dtype to ScalarType."""
+        """Convert a torch dtype to JitScalarType.
+
+        NB: DO NOT USE this API if dtype is comes from a torch._C.Value.type().dtype() call because
+            "RuntimeError: INTERNAL ASSERT FAILED at "../aten/src/ATen/core/jit_type_base.h" can
+            be raised in some scenarios. Instead use from_value API which handles all cases
+
+        Args:
+            dtype: A torch.dtype to create a JitScalarType from
+
+        Returns:
+            JitScalarType
+
+        Raises:
+            ValueError: if dtype is not a valid torch.dtype or if it is None.
+        """
         if dtype not in _DTYPE_TO_SCALAR_TYPE:
             raise ValueError(f"Unknown dtype: {dtype}")
         return _DTYPE_TO_SCALAR_TYPE[dtype]
 
+    # TODO: `value: torch._C.Value` type annotation raises (CI machines only):
+    #       NameError: name '_C' is not defined
+    # In both cases, using typing.TypeVar or string literals didn't work
+    @classmethod
+    @_beartype.beartype
+    def from_value(cls, value) -> JitScalarType:
+        """Create a JitScalarType from a torch._C.Value underlying scalar type.
+
+        Args:
+            value: A `torch._C.Value` object to fetch scalar type from.
+
+        Returns:
+            JitScalarType.
+
+        Raises:
+            OnnxExporterError: if value is None.
+            SymbolicValueError: when value.type().scalarType() does not exist
+
+        """
+
+        if value is None:
+            raise errors.OnnxExporterError(
+                "Cannot determine scalar type for `None` instance."
+            )
+        elif isinstance(value, torch.Tensor):
+            return JitScalarType.from_dtype(value.dtype)
+        elif isinstance(value.type(), torch.ListType):
```

Comment:
nit: value.type() is called many times so maybe store it in a variable? No strong opinion though

Reply:
ok, will leave as it is